pkg/util/vsphere/vm/cdrom.go

```diff
@@ -0,0 +1,388 @@
+		Device:    ahciController,
+		Operation: vimtypes.VirtualDeviceConfigSpecOperationAdd,
+	})
+	curDevices = append(curDevices, ahciController)
+
+	return ahciController, deviceChanges, curDevices
+}
+
+// getIsoFilenameFromImageRef returns the ISO type content library file name
+// based on the given VirtualMachineImageRef.
+// If a namespace scope VirtualMachineImage is provided, it checks the
+// ContentLibraryItem status, otherwise, a ClusterVirtualMachineImage status,
+// and gets the ISO file storage URI with the server ID suffix dropped.
+func getIsoFilenameFromImageRef(
+	vmCtx pkgctx.VirtualMachineContext,
+	client ctrlclient.Client,
+	imageRef vmopv1.VirtualMachineImageRef) (string, error) {
+	var itemStatus imgregv1a1.ContentLibraryItemStatus
+
+	switch imageRef.Kind {
+	case vmiKind:
+		ns := vmCtx.VM.Namespace
+		var vmi vmopv1.VirtualMachineImage
+		if err := client.Get(vmCtx, ctrlclient.ObjectKey{Name: imageRef.Name, Namespace: ns}, &vmi); err != nil {
+			return "", err
+		}
+		if vmi.Spec.ProviderRef == nil {
+			return "", fmt.Errorf("provider ref is nil for VirtualMachineImage: %q", vmi.Name)
+		}
+		var clitem imgregv1a1.ContentLibraryItem
+		if err := client.Get(vmCtx, ctrlclient.ObjectKey{Name: vmi.Spec.ProviderRef.Name, Namespace: ns}, &clitem); err != nil {
+			return "", err
+		}
+		itemStatus = clitem.Status
+	case cvmiKind:
+		var cvmi vmopv1.ClusterVirtualMachineImage
+		if err := client.Get(vmCtx, ctrlclient.ObjectKey{Name: imageRef.Name}, &cvmi); err != nil {
+			return "", err
+		}
+		if cvmi.Spec.ProviderRef == nil {
+			return "", fmt.Errorf("provider ref is nil for ClusterVirtualMachineImage: %q", cvmi.Name)
+		}
+		var cclitem imgregv1a1.ClusterContentLibraryItem
+		if err := client.Get(vmCtx, ctrlclient.ObjectKey{Name: cvmi.Spec.ProviderRef.Name}, &cclitem); err != nil {
+			return "", err
+		}
+		itemStatus = cclitem.Status
+
+	default:
+		return "", fmt.Errorf("unsupported image kind: %q", imageRef.Kind)
+	}
+
+	if itemStatus.Type != imgregv1a1.ContentLibraryItemTypeIso {
+		return "", fmt.Errorf("expected ISO type image, got %s", itemStatus.Type)
+	}
+	if len(itemStatus.FileInfo) == 0 || itemStatus.FileInfo[0].StorageURI == "" {
+		return "", fmt.Errorf("no storage URI found in the content library item status: %v", itemStatus)
+	}
+
+	// Currently the storage URI is in the following format:
+	// "ds:///vmfs/volumes/...{file_id}.iso?serverId=..."
+	//
+	// vCenter converts this to a different format during the CD-ROM creation:
+	// "[sharedVmfs-0] contentlib-${lib_id}/${item_id}/...{file_id}.iso"
+	//
+	// Therefore, we should return the storage URI in the vCenter format to
+	// ensure the existing CD-ROMs can be fetched by the backing file name.
+	//
+	// TODO(Sai): Update image-registry-operator to put storageURI in the
+	// vCenter format in content library item CR status and remove below lines.
+	uri := itemStatus.FileInfo[0].StorageURI
+	if idex := strings.Index(uri, serverIDSuffix); idex != -1 {
```

Comment:
This is not needed anymore since we have the internal image-registry-operator updated with the expected format of URI stored in the content library item's CR status.  

Thanks for providing the example. I will keep that in mind for future similar changes.